dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+#include "dynamic_array.h"
+
+// NOTE: When we destroy items, we call the destructor for them 
+// iff the destructor is not NULL.
+
+typedef void (*destr)(void *);
+
+// Initialiser - sets fields in instance and allocates buffer (via buffer_alloc)
+void buffer_init(struct buffer *inst, size_t item_size, size_t capacity, 
+                 size_t alloc_by, void (*destructor)(void *))
+{   
+    inst->capacity   = 0;
```

Comment:
For zero-ing out a structure, we can also use `memset(ptr, 0, bytes)`.  Zeroing out the entire structure then setting a few fields is sometimes quicker than zeroing fields individually, if most fields are to be zero.  Speed is not too important here but it is also safer since it guarantees that all fields are zero-initialised when we initialise the structure (e.g. if more fields were added later but forgot to be explicitly zeroed in initialiser).

I'd recommend that we `memset` the `inst` to zero: `memset(inst, 0, sizeof(*inst));` then set the non-zero fields item_size, alloc_by, etc.